storage capacity GA

pohly opened the pull request:
#### What type of PR is this?

/kind feature
/kind api-change

#### What this PR does / why we need it:

[Storage capacity tracking](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1472-storage-capacity-tracking) is ready for promotion to GA.

#### Which issue(s) this PR fixes:
Related-to: https://github.com/kubernetes/enhancements/issues/1472

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
CSIStorageCapacity.storage.k8s.io: The v1beta1 version of this API is deprecated in favor of v1, and will be removed in v1.27. If a CSI driver supports storage capacity tracking, then it must get deployed with a release of external-provisioner that supports the v1 API.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/1472
- [Usage]: https://kubernetes.io/docs/concepts/storage/storage-capacity/
- [Other doc]: https://kubernetes-csi.github.io/docs/storage-capacity-tracking.html
```

pohly committed:
storage capacity: GA, always enabled, remove feature check

The code and tests for scenarios where the feature is disabled are no longer
needed because the feature is graduating to GA.

pohly committed:
storage capacity: v1 API

pohly committed:
storage capacity: generated files

pohly committed:
storage capacity: use V1 API

liggitt commented on test/integration/etcd/data.go:
```diff
@@ -327,6 +327,10 @@ func GetEtcdStorageDataForNamespace(namespace string) map[schema.GroupVersionRes
 		},
+		gvr("storage.k8s.io", "v1", "csistoragecapacities"): {
+			Stub:             `{"metadata": {"name": "csc-12345-3"}, "storageClassName": "sc1"}`,
+			ExpectedEtcdPath: "/registry/csistoragecapacities/" + namespace + "/csc-12345-3",
```
this should expect v1beta1 still:
```
ExpectedGVK:      gvkP("storage.k8s.io", "v1beta1", "CSIStorageCapacity"),
```

pohly replied:
That worked, thanks.

sftim commented:
Could we defer the removal of the v1beta1 API for some extra releases? There will surely be some CSI drivers that are slow to update. I suggest removing that deprecated v1beta1 API in v1.28 - is that reasonable?

:thought_balloon: do we perhaps have a mechanism to achieve this:
| Minor version  | Details |
| ------------- | ------------- |
| v1.24  | v1beta1 available but deprecated  |
| v1.25  |  v1beta1 available but deprecated  |
| v1.26  |  v1beta1 available but deprecated  |
| v1.27  |  v1beta1 deprecated and off by default. You can enable it manually.  |
| v1.28  |  v1beta1 fully removed |

pohly commented:
> Could we defer the removal of the v1beta1 API for some extra releases? There will surely be some CSI drivers that are slow to update.

Only those which use the feature need to update.

> I suggest removing that deprecated v1beta1 API in v1.28 - is that reasonable?

I suspect that the answer will be "no, because we want people to stop depending on beta APIs", but I'll let @liggitt provide the authoritative answer for this.

pohly commented:
/retest

Timeout after 2 hours in pull-kubernetes-integration.

liggitt commented:
> 

Off by default in 1.27 is the key.

There is actually already a mechanism to achieve the one-release re-enablement grace period that would let it be turned back on in 1.28 only; setting this envvar when running `kube-apiserver`:

```
KUBE_APISERVER_SERVE_REMOVED_APIS_FOR_ONE_RELEASE=true
```

I don't think it's currently documented, as it was intended as an escape hatch and not a common practice.

sftim commented:
> `KUBE_APISERVER_SERVE_REMOVED_APIS_FOR_ONE_RELEASE`
Today I Learned!

leilajal commented:
/remove-sig api-machinery

leilajal commented:
/remove-sig api-machinery

fedebongio commented:
/remove-sig api-machinery

alculquicondor commented:
/assign @msau42 @alculquicondor 

for sig-storage and sig-scheduling, respectively

roycaihw reviewed (commented):
API review looks good otherwise

I didn't review the `pkg/scheduler`, `pkg/printers` and `plugin/pkg/auth` which changes how they used the feature gate

pohly commented on staging/src/k8s.io/api/storage/v1beta1/types.go:
```diff
@@ -364,11 +364,7 @@ type CSIDriverSpec struct {
 	//
 	// This field was immutable in Kubernetes <= 1.22 and now is mutable.
 	//
-	// This is a beta field and only available when the CSIStorageCapacity
```
Good catch, I missed that one.

While editing I also noticed that I should have used CSIDriverSpec.StorageCapacity instead of CSIDriver.StorageCapacity - also fixed.

pohly commented on staging/src/k8s.io/api/storage/v1/types.go:
```diff
@@ -560,3 +561,101 @@ type CSINodeList struct {
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:prerelease-lifecycle-gen:introduced=1.24
```
Removed.

pohly commented on pkg/features/kube_features.go:
```diff
@@ -157,6 +157,7 @@ const (
 	// owner: @pohly
 	// alpha: v1.19
 	// beta: v1.21
+	// GA: v1.24
```
Will do.

pohly commented on pkg/features/kube_features.go:
```diff
@@ -864,7 +865,7 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 	InTreePluginPortworxUnregister:                 {Default: false, PreRelease: featuregate.Alpha},
 	CSIInlineVolume:                                {Default: true, PreRelease: featuregate.Beta},
-	CSIStorageCapacity:                             {Default: true, PreRelease: featuregate.Beta},
+	CSIStorageCapacity:                             {Default: true, PreRelease: featuregate.GA, LockToDefault: true}, // remove in 1.26
```
The design is done and implemented and we have given developers as much time as we can before graduation.

Usage numbers turned out to be harder to determine with certainty. I know of people using it who didn't reply when I asked them to report back.

alculquicondor commented:
/hold
for update in KEP about Autoscaling details

alculquicondor commented:
To be clear, I'm not asking for a CA implementation, just a clear path for it to be implemented in the future.

pohly commented:
> hold for update in KEP about Autoscaling details

I'll add those.

> To be clear, I'm not asking for a CA implementation, just a clear path for it to be implemented in the future.

There is also already an implementation which I used successfully for scale up from zero and scaling down: https://github.com/kubernetes/autoscaler/pull/3887

msau42 commented on staging/src/k8s.io/api/storage/v1/types.go:
```diff
@@ -560,3 +558,99 @@ type CSINodeList struct {
+	// GetCapacityRequest. The corresponding value in the Kubernetes
+	// API is ResourceRequirements.Requests in a volume claim.
+	//
+	// +optional
```
What happens if unset?

pohly replied:
`Capacity` is used as an approximation.

Both fields not set [is treated](https://github.com/kubernetes/kubernetes/blob/debd461d03b0643ebcc34dc818c2f274d11b8817/pkg/scheduler/framework/plugins/volumebinding/binder.go#L976-L983) as "no capacity". This cannot occur in practice right now because external-provisioner doesn't generate such objects.

alculquicondor commented on staging/src/k8s.io/api/storage/v1/types.go:
```diff
@@ -560,3 +558,99 @@ type CSINodeList struct {
+	// The semantic is currently (CSI spec 1.2) defined as:
+	// The available capacity, in bytes, of the storage that can be used
+	// to provision volumes. If not set, that information is currently
+	// unavailable and treated like zero capacity.
```
Unless MaximumVolumeSize is defined?

pohly replied:
If MaximumVolumeSize is set and Capacity is not, then MaximumVolumeSize will be used.

This doesn't occur in practice right now. Because CSI drivers always report `available_capacity`.

alculquicondor replied:
Document what is allowed in the comment, and update the validation accordingly.

pohly replied:
> Document what is allowed in the comment

The documentation is correct: both fields are optional and the creator decides what it provides.

What I can clarify is the interpretation of "MaximumVolumeSize non-nil and Capacity nil": as implemented and IMHO implied, the scheduler then uses "MaximumVolumeSize".

pohly replied:
@alculquicondor Let's discuss the revised comments in https://github.com/kubernetes/enhancements/pull/3233 first, see https://github.com/kubernetes/enhancements/pull/3233/commits/c1505fef482e0e38f78e3d4e0b58e751133b40bd.

pohly replied:
Then I'll copy them into the various types.go files and update the generated files.

pohly replied:
I've updated the comments as agreed in https://github.com/kubernetes/enhancements/pull/3233#discussion_r821853048.

leilajal commented:
/remove-sig api-machinery

pohly committed:
storage capacity: clarify relationship between capacity fields

pohly commented:
/retest

fedebongio commented:
/triage accepted
/remove-sig api-machinery

pohly committed:
storage capacity: CRUD conformance test

This is needed for GA APIs. This one is based on
test/e2e/common/node/runtimeclass.go.

liggitt commented on test/e2e/storage/csistoragecapacity.go:
```diff
@@ -0,0 +1,209 @@
+	framework.ConformanceIt(" should support CSIStorageCapacities API operations", func() {
+		// Setup
+		cscVersion := "v1"
+		cscClient := f.ClientSet.StorageV1().CSIStorageCapacities("default")
```
this should use the test's namespace, not the default namespace, right?

pohly replied:
Right, fixed.

liggitt commented on test/e2e/storage/csistoragecapacity.go:
```diff
@@ -0,0 +1,209 @@
+		ginkgo.By("deleting")
+		err = cscClient.Delete(context.TODO(), createdCSC.Name, metav1.DeleteOptions{})
+		framework.ExpectNoError(err)
+		_, err = cscClient.Get(context.TODO(), createdCSC.Name, metav1.GetOptions{})
+		framework.ExpectEqual(apierrors.IsNotFound(err), true, fmt.Sprintf("expected 404, got %#v", err))
```
this will error/flake if a finalizer was added to the object... we shouldn't prevent a cluster from being conformant if they add finalizer behavior on storagecapacities. make sure it either gets a 404 or has a deletionTimestamp set with a finalizer

see the `// If ingress controller does not support finalizers, we expect a 404.  Otherwise we validate finalizer behavior.` bit in the ingress conformance test

liggitt commented:
API promotion lgtm. Defer review of the scheduler/storage changes to scheduler/storage teams.

Conformance test is only exercising API CRUD, not behavior, which seems appropriate for a generic test that doesn't know what specific CSI drivers are installed in a cluster. Left a few comments on the conformance test.

liggitt commented on test/e2e/storage/csistoragecapacity.go:
```diff
@@ -0,0 +1,209 @@
+		framework.ExpectEqual(gottenCSC.UID, createdCSC.UID)
+
+		ginkgo.By("listing")
+		cscs, err := cscClient.List(context.TODO(), metav1.ListOptions{LabelSelector: "test=" + f.UniqueName})
```
should also list/watch across namespaces to exercise those two endpoints

pohly replied:
Done.

I've force-pushed to keep this in a mergeable state once it has LGTM, here are the changes: https://github.com/kubernetes/kubernetes/compare/65aee2038850a980500f1ccb25cf5dfc781748e7..ff47fb98735caf3578b7720cc806d2d2c195d684

liggitt commented:
another question for @msau42 ... is the behavior of this exercised in a non-conformance e2e with a mock CSI driver? if not, should it be?

pohly commented:
> is the behavior of this exercised in a non-conformance e2e with a mock CSI driver?

Yes, it is: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/capacity.go#L111

pohly commented on test/e2e/storage/csistoragecapacity.go:
```diff
@@ -0,0 +1,209 @@
+		Description:
+		The storage.k8s.io API group MUST exist in the /apis discovery document.
+		The storage.k8s.io/v1 API group/version MUST exist in the /apis/mode.k8s.io discovery document.
+		The runtimeclasses resource MUST exist in the /apis/storage.k8s.io/v1 discovery document.
+		The runtimeclasses resource must support create, get, list, watch, update, patch, delete, and deletecollection.
```
Ooh, right. Fixed.

pohly commented on test/e2e/storage/csistoragecapacity.go:
```diff
@@ -0,0 +1,209 @@
+		framework.ExpectNoError(err)
+		framework.ExpectEqual(len(cscs.Items), 2, "filtered list should have 2 items")
+
+		ginkgo.By("deleting a collection")
```
Done.

msau42 commented:
storage bits LGTM
/approve

pohly commented:
@liggitt: all tests are green now. KEP update was merged. Can I get the final LGTM + root approval?

liggitt commented on test/e2e/storage/csistoragecapacity.go:
```diff
@@ -0,0 +1,252 @@
+		}
+		cscs, err = cscClient.List(context.TODO(), metav1.ListOptions{LabelSelector: "test=" + f.UniqueName})
+		framework.ExpectNoError(err)
+		framework.ExpectEqual(len(cscs.Items), remaining, "filtered list should have %d items", remaining)
```
the finalizer observed at delete time could be handled and removed by the time we list. should be <= remaining, I think

pohly replied:
True. We also have 2 as lower bound as only one object was deleted. I've updated the check and also fixed some other incomplete copy-and-paste (What is a CSIStorageClass? A hybrid of RuntimeClass and CSIStorageCapacity...) -> https://github.com/kubernetes/kubernetes/compare/e86db9d8dba39716c6ddb5bf5519119990d85f24..b54bfe93fffb9d4f6b958e4821d05874c4742790

liggitt commented:
/approve
for API changes

one comment on the list-after-delete expected count in the test

conformance yaml update needs approval from someone in that group

pohly commented:
/assign @dims 

For conformance approval.

alculquicondor reviewed (commented):
/approve
pkg/scheduler

pohly commented:
/assign @smarterclayton

For test/conformance/testdata approval.

pohly commented:
/retest

hh commented on test/conformance/testdata/conformance.yaml:
```diff
@@ -2339,6 +2339,16 @@
+    discovery document. The csistoragecapacities resource must support create, get,
+    list, watch, update, patch, delete, and deletecollection.'
+  release: v1.24
+  file: test/e2e/storage/csistoragecapacity.go
```
Let's make sure this test has been green on test grid for at least two weeks, before we upgrade it to `[Conformance]`.

hh commented on test/e2e/storage/csistoragecapacity.go:
```diff
@@ -0,0 +1,257 @@
+		The csistoragecapacities resource MUST exist in the /apis/storage.k8s.io/v1 discovery document.
+		The csistoragecapacities resource must support create, get, list, watch, update, patch, delete, and deletecollection.
+	*/
+	framework.ConformanceIt(" should support CSIStorageCapacities API operations", func() {
```
Ideally this test is submitted without 'ConformanceIt' for a couple weeks, then when it's shown stable, we add `ConformanceIt()`, updated the `conformance/testdata`. Is that possible?

pohly replied:
Sure. I've changed this test so that it gets added as a normal test.

https://github.com/kubernetes/kubernetes/pull/108867 has the promotion to conformance.

pohly replied:
For example, the test runs in `pull-kubernetes-e2e-kind` and should be in any of the periodic storage tests:
```
[It]  should support CSIStorageCapacities API operations
  test/e2e/storage/csistoragecapacity.go:48
[1mSTEP[0m: getting /apis
[1mSTEP[0m: getting /apis/storage.k8s.io
[1mSTEP[0m: getting /apis/storage.k8s.io/v1
[1mSTEP[0m: creating
[1mSTEP[0m: watching
Mar 22 07:53:29.629: INFO: starting watch
[1mSTEP[0m: getting
[1mSTEP[0m: listing in namespace
[1mSTEP[0m: listing across namespaces
[1mSTEP[0m: patching
[1mSTEP[0m: updating
Mar 22 07:53:29.736: INFO: waiting for watch events with expected annotations in namespace
Mar 22 07:53:29.736: INFO: waiting for watch events with expected annotations across namespace
[1mSTEP[0m: deleting
[1mSTEP[0m: deleting a collection
[AfterEach] [sig-storage] CSIStorageCapacity
  test/e2e/framework/framework.go:186
Mar 22 07:53:29.832: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
[1mSTEP[0m: Destroying namespace "csistoragecapacity-5900" for this suite.

[32m•[0m
[90m------------------------------[0m
{"msg":"PASSED [sig-storage] CSIStorageCapacity  should support CSIStorageCapacities API operations","total":-1,"completed":30,"skipped":286,"failed":0}
```

xing-yang commented:
/retest

xing-yang commented:
/lgtm
/approve

pohly commented:
/retest

pohly commented:
/retest

hh commented on test/e2e/storage/csistoragecapacity.go:
```diff
@@ -0,0 +1,258 @@
+		The csistoragecapacities resource must support create, get, list, watch, update, patch, delete, and deletecollection.
+		TODO (pohly): promote to framework.ConformanceIt
+	*/
+	ginkgo.It(" should support CSIStorageCapacities API operations", func() {
```
After a couple weeks of non-flaky soaking, we can merge in the conformance promotion PR https://github.com/kubernetes/kubernetes/pull/108867/files#diff-5e1072eb7b333bc3faf7068eb5ce93d29adaad658b5bcdd52a8238c18c5f8f92R47

pohly commented:
/retest

Riaankl commented:
@pohly the test grid for the test have been green every time I checked. :smile: 
Test freeze is on 6 April. Can you please create the Promtion PR before then to be sure no "new  conformance tech debt" will appear for this release.
